sway-lib-std/src/u256.sw

```diff
@@ -395,21 +395,58 @@ impl core::ops::Subtract for U256 {
 impl core::ops::Multiply for U256 {
     /// Multiply a `U256` with a `U256`. Panics on overflow.
     fn multiply(self, other: Self) -> Self {
-        let zero = U256::from((0, 0, 0, 0));
-        let one = U256::from((0, 0, 0, 1));
-
-        let mut x = self;
-        let mut y = other;
-        let mut result = U256::new();
-        while y != zero {
-            if (y & one).d != 0 {
-                result += x;
+        // both of the upper words cannot be nonzero
+        // otherwise overflow is automatic
+        assert(self.a == 0 || other.a == 0);
+
+        if self.a != 0 {
+            // in case self.a is nonzero, all of the
+            // words of other, except for d, should be 
+            // zero otherwise overflow is automatic
+            assert(other.b == 0 && other.c == 0);
+            U256::from((self.a * other.d, 0, 0, 0))
+        } else if other.a != 0 {
+            // in case other.a is nonzero, all of the
+            // words of self, except for d, should be 
+            // zero otherwise overflow is automatic
+            assert(self.b == 0 && self.c == 0);
+            U256::from((other.a * self.d, 0, 0, 0))
+        } else {
+            if self.b != 0 {
+                // in case self.b is nonzero, other.b has 
+                // to be zero, otherwise overflow is automatic
+                // due to the fact that other.b * 2 ^ (64 * 2) *
+                // self.b * 2 ^ (62 ^ 2) > 2 ^ (64 * 4)
+                assert(other.b == 0);
+                let result_b_d = self.b.overflowing_mul(other.d);
+                let result_c_c = self.c.overflowing_mul(other.d);
+                let result_c_d = self.c.overflowing_mul(other.d);
+                let result_d_c = self.d.overflowing_mul(other.c);
+                let result_d_d = self.d.overflowing_mul(other.d);
+
+                U256::from((self.b * other.c + result_b_d.upper, result_b_d.lower + result_c_d.upper + result_d_c.upper, result_d_d.upper + result_c_d.lower + result_d_c.lower, result_d_d.lower))
```

Comment:
I would change these longer lines to use variables e.g. 

`U256::from(a, b, c, d)` 

where 

`let a = self.b * other.c + result_b_d.upper`.

Also, why are the values inside the `from()` wrapped in a tuple i.e. it looks like `from( (values ...) )`?

Reply:
Now `from` function from `From` trait takes only one argument - the tuple

Reply:
Yeah we had to change that recently in order to unify the definition of `from`, just like Rust does.